refactor: use interface for auth config object

janriemer opened the pull request:
This makes use of the IAuthConfig interface
for the config object that is passed into
the ctor of AuthConfig class and returned
by getConfig method. Furthermore it simplifies
assignment of config object.

Unfortunately, I was **unable to run the tests** with `npm test`, cause i got the follow error from karma (stackoverflow question concerning the same error): http://stackoverflow.com/questions/24220888/error-you-need-to-include-some-adapter-that-implements-karma-start-method#24221029

Based on answers from stackoverflow, it seems that there is a karma.conf.js file missing. I have not used karma before, so at the moment I can't help you with that.

escardin commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,41 @@ export interface IAuthConfig {
-      this.headerPrefix = '';
+  private _config : IAuthConfig;
+
+  constructor(config:IAuthConfig = {
```
I'm pretty sure this won't work if the user only specifies some config and not all of it.

janriemer replied:
Hmm... you are probably right. Thanks for catching that. :rose: 
And making all props of IAuthConfig optional is not a viable solution either, I guess!?

What do you think of introducing another interface (like `IAuthConfigOptional`) whose
properties are all optional? This would then only be used as an argument for the ctor.
Of course we have to cast, when we assign it to `IAuthConfig`.
In order to clarify, here is an example:
```typescript
interface IConfig {
	foo: Number;
	bar: string;
}

interface IConfigOptional {
	foo?: Number;
	bar?: string;
}

class FooBar {
	_config: IConfig;
	
	constructor(config: IConfigOptional = {
		foo: 2,
		bar: "baz"
	}){
		// we have to cast here, but that is no problem,
		// as long as we have default values for all props
		// in ctor definition
		this._config = <IConfig>config;
	}
	
	
}

// we do not specify all props here
var test = new FooBar({foo:2});

escardin replied:
Way too complicated. We can just do this:

```
const defaultConfig={
      headerName: AuthConfigConsts.DEFAULT_HEADER_NAME,
      headerPrefix: null,
      tokenName: AuthConfigConsts.DEFAULT_TOKEN_NAME,
      tokenGetter: (() => localStorage.getItem(AuthConfigConsts.DEFAULT_TOKEN_NAME)),
      noJwtError: false,
      globalHeaders: [],
      noTokenScheme: false
  }
this._config = Object.assign(defaultConfig, config);
```

janriemer replied:
I don't see what's complicated about it. It is just one more interface... :confused: 

Of course your solution works, too, but it fails the point I'm trying to achieve with using interfaces, which is **type information**. I mean, we are using typescript here, so why not using the type info where possible?
The signature `AuthHttp(config:any)` does not say _anything_ about the form of the config object, so IMHO type info would significantly increase usability.

Do you have other ideas of providing type info to the config object?

escardin replied:
You can still have typings for the interface. You just make all the fields optional.

The AuthConfig API might benefit from a more conscious refactoring than just adding typings though. It's essentially a wrapper around a config object and offers minimal functionality on its own.

janriemer replied:
> You can still have typings for the interface. You just make all the fields optional.

That's what I have proposed in the first paragraph of [this comment](https://github.com/auth0/angular2-jwt/pull/93#discussion_r65080928), but I think we will get in trouble having all these optional fields further down the road, because when calling `getConfig` method, we know that all properties are set and they are not optional anymore.

escardin replied:
Ah, yes two interfaces. I'm just blind today and not at my best. Apologies. Two interfaces would work.

escardin commented:
Regarding the tests. The repo ignores all .js files (of which my karma config is one) and so I didn't realise that wasn't comitted.

janriemer commented:
OK, I've updated the PR. What do you think?

escardin commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,62 @@ export interface IAuthConfig {
+    noTokenScheme?: boolean;
+}
+
+class AuthConfigConsts {
```
I don't really see the point of putting this in a class if it's not exported

janriemer replied:
How would you do it then?
I think it is more maintainable, if we have them under one namespace.

escardin replied:
Just have them. They don't need a namespace (technically they're part of the angular2-jwt module). For the most part, they're used exactly once in the default config, so factoring them out isn't a real benefit either. It makes sense to export them so that users can reference them for their own purposes.

(While I think things like SOLID are excellent ideas and practices, they're guidelines, not dogma to be worshipped).

janriemer replied:
OK, now the class with the constants is exported.

escardin commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,62 @@ export interface IAuthConfig {
+    public static readonly HEADER_PREFIX_BEARER = 'Bearer ';
+}
+
+export const AuthConfigDefaults: IAuthConfig = {
```
I don't really think this should be exported

janriemer replied:
OK, I've updated it.

escardin commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,62 @@ export interface IAuthConfig {
-      this.headerPrefix = '';
+  private _config: IAuthConfig;
+
+  constructor(config: IAuthConfigOptional = {
```
This still forces someone who only wants to provide one or two options to provide all of them.

escardin commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,54 @@ export interface IAuthConfig {
-      this.headerPrefix = '';
+  private _config: IAuthConfig;
+
+  constructor(config: IAuthConfigOptional = AuthConfigDefaults) {
```
Like this:
```
  constructor(config: IAuthConfigOptional = {}) {
    this._config = Object.assign({}, AuthConfigDefaults, config);
```

janriemer replied:
I don't see the point in doing it the way you propose.
The signature of the ctor should tell the client what the shape of the config object is.
In your solution one could assume, that the default value is an empty object, which is not the case.

escardin replied:
Fixed. The point is that the typing tells you what you can provide. If you don't provide anything, we'll use the defaults. What your default does is only set the values that are passed and leave the ones that aren't blank. If you provide nothing, they are equivalent.

janriemer replied:
Thanks for catching that. I've updated the PR.
I wonder, though, why it is not possible in typescript to express it the way I've done it.

For me it was just an optional object, which has itself optional fields with default values.
When I specify some of these fields, but not all, I would expect that the other fields use their default values, **but this is not the case**, as you @escardin have correctly pointed out.
I find this behavior really counterintuitive, but anyway... :confused:

escardin commented:
> Thanks for catching that. I've updated the PR.
> I wonder, though, why it is not possible in typescript to express it the way I've done it.
> For me it was just an optional object, which has itself optional fields with default values.
> When I specify some of these fields, but not all, I would expect that the other fields use their default values, but this is not the case, as you @escardin have correctly pointed out.
> I find this behavior really counterintuitive, but anyway... 😕 


The reason for that is that the entire object is the default argument, not the properties within it. If you look at what typescript generates, I bet it looks like `blah=blah||{/*complex object*/}`. To do it the way you've done it would require additional inspection of the object and I believe is outside the es6 spec.

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Default_parameters

chenkie commented:
Not too sure where we're at here, is this ready to be merged? :)

janriemer commented:
Hi @chenkie,
thanks for asking. In my opinion it can be merged now.

janriemer commented:
Please tell me what is still missing, so that this PR finally gets merged.

chenkie commented:
@janriemer sorry for the uber delay. I merged a PR that dealt with global headers which introduces some conflict for yours. Can you take a look for it and make any necessary changes? Appreciate your PR :)

janriemer commented:
@chenkie No problem.
All righty, I've updated it. :wink:

chenkie commented:
Thanks! I'm getting some issues with `readonly` on the `AuthConfigConsts` though.

```
angular2-jwt.ts(29,28): error TS1005: '=' expected.
angular2-jwt.ts(30,28): error TS1005: '=' expected.
angular2-jwt.ts(31,28): error TS1005: '=' expected.
```

Are you getting that?

janriemer commented:
I'm sorry for that. Thanks for pointing it out.

I don't get these errors in my editor (using atom with [atom-typescript](https://github.com/TypeStrong/atom-typescript)).
Atom-Typescript uses the nightly builds of typescript, so it seems `readonly` has not made it into the stable version yet.
Anyway, I've updated the PR by removing `readonly`. It is not an ideal solution though, because now fields that are supposed to be read only can easily be reassigned.
What do you think?

chenkie commented:
We can probably just leave the `readonly` out for now and put it back in when it gets shipped in Typescript.

The only issue I'm running into is that changing the `tokenName` in the config doesn't affect the `tokenGetter`, so the default will always be applied. How can we adjust to allow it to take the optional `tokenName`?

janriemer commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,54 @@ export interface IAuthConfig {
+    headerName: AuthConfigConsts.DEFAULT_HEADER_NAME,
+    headerPrefix: null,
+    tokenName: AuthConfigConsts.DEFAULT_TOKEN_NAME,
+    tokenGetter: () => localStorage.getItem(AuthConfigDefaults.tokenName),
```
Sorry, I don't know how to run the tests. They would have easily caught that.

tokenGetter now uses tokenName of the AuthConfigDefaults object.

janriemer commented on angular2-jwt.ts:
```diff
@@ -15,46 +15,54 @@ export interface IAuthConfig {
+
+  constructor(config?: IAuthConfigOptional) {
+    config = config || {};
+    this._config = Object.assign({}, AuthConfigDefaults, config);
```
We will get into trouble when using `Object.assign` and targeting `ES5` (see [compat table](https://kangax.github.io/compat-table/es6/#test-Object_static_methods_Object.assign)). There is a [ponyfill](https://github.com/sindresorhus/object-assign) out there, though. Is this an option? It seems to be pretty well maintained.
So before this issue is not resolved, we **should not merge** this.
Any comments @chenkie?

chenkie replied:
The ponyfill looks pretty good. Can you give it a shot and let me know if it works out?

janriemer commented:
Hi @chenkie,
the object-assign ponyfill is now implemented.
I've also added typings for the param in `provideAuth` function.

I've tested the ponyfill in my own little test.ts file and it works like a charm. :+1: 

However, I'm not able to test angular2-jwt as a whole. I've tried to set up karma (with karma.conf.js file), but always get an error (`You need to include some adapter that implements __karma__.start method!`). We should definitely invest some time in correctly setting up the tests.

So IMHO this can be merged now, unless we want safety-first and have tests correctly set up.

This fixes #27.

chenkie commented:
@janriemer checking this out now :)

chenkie commented:
Did you run into any issues with requiring object-assign when you were testing? Trying to test it out in an app but I'm getting 404's for it.

janriemer commented:
Sorry, I'm quite busy right now. I'll have a deeper look at it in the end of August. :fallen_leaf:

janriemer commented:
Hey @chenkie,
I've updated this PR again and rebased it onto current master.
I saw that we now also have a karma.conf. Really nice! :sparkles: :tada: 

I've run the tests and **they are all passing**. :rocket:
However, I get **compile errors in almost all d.ts files** (like jasmine.d.ts etc):
It complains about `duplicate identifiers`. Don't know what's wrong there :confused:,
but apparently it **doesn't affect the actual library**.

IMHO, this can be merged now.

janriemer commented:
Hey @chenkie,
are there any issues left?
I would love to see this being merged. :sparkles:

chenkie commented:
I'm still getting some errors when I try to build it

```bash
typings/browser/ambient/es6-shim/index.d.ts(8,14): error TS2300: Duplicate identifier 'PropertyKey'.
typings/browser/definitions/object-assign/index.d.ts(11,1): error TS2300: Duplicate identifier 'export='.
typings/browser/definitions/object-assign/index.d.ts(15,1): error TS2300: Duplicate identifier 'export='.
typings/browser/definitions/object-assign/index.d.ts(19,1): error TS2300: Duplicate identifier 'export='.
typings/main/ambient/es6-shim/index.d.ts(8,14): error TS2300: Duplicate identifier 'PropertyKey'.
typings/main/ambient/jasmine/index.d.ts(82,9): error TS2375: Duplicate number index signature.
typings/main/ambient/jasmine/index.d.ts(136,9): error TS2374: Duplicate string index signature.
typings/main/definitions/object-assign/index.d.ts(11,1): error TS2300: Duplicate identifier 'export='.
typings/main/definitions/object-assign/index.d.ts(15,1): error TS2300: Duplicate identifier 'export='.
typings/main/definitions/object-assign/index.d.ts(19,1): error TS2300: Duplicate identifier 'export='.
```

janriemer commented:
These errors are really annoying. As I stated [above](https://github.com/auth0/angular2-jwt/pull/93#issuecomment-244375100), I don't know what causes them.
The error `typings/browser/ambient/es6-shim/index.d.ts(8,14): error TS2300: Duplicate identifier 'PropertyKey'.` must have been present before this PR, because I haven't touched es6-shim at all.
Also, definitions of jasmine seems to have errors, which I haven't touched either.

However, I've updated the object-assign definitions now (like 5 min ago) and make it `ambient`.

janriemer commented:
OK, I've looked into it again and I have no idea what the issue is.
Of course the **only problem are the type definitions for object assign** and this leads into a cascade of errors in the tsc. But why?
Sorry for bothering you. It seems this PR still has to wait.

Has someone any idea on how we can work around this issue?

What really bothers me is the fact that it runs and works (all tests pass) and even the autocompletion for object-assign are working in my editor (Atom with atom-typescript plugin).
The only thing that gets in our way at the moment is the tsc. I thought it should make us more productive and not hinder us. :confused:

chenkie commented:
yeah that's frustrating 😞 we'll get it sorted though.

@escardin do you have any ideas?

escardin commented:
Looking into it now.

escardin commented:
Peer deps are also outdated.

escardin commented:
If the use @types pr gets merged first, you'll need to add     "@types/object-assign": "^4.0.28",
to this branches dev-dependencies

janriemer commented:
Hey @escardin,
Awesome, thanks for your help! I've updated the PR accordingly (and hope for the best :pray: ).
Tsc seems to be happy and tests pass without errors.

However, I get an error with atom-typescript saying that it can't find name `Promise`.
Could the reason be the now missing browser.d.ts file?
I am so confused right now! :confused: 

Thank you @chenkie and @escardin for your patience. :rose:

escardin commented:
Typescript 2.0 has a lot of quality of life improvments surrounding typings. The first is that typings are looked for implicitly, which makes most of what we had in our typings.json obsolete (we use @types/jasmine instead of the dt in typings). The other major part is that TS2.0 has more built in support for es6 and similar, which needs to be explicitly supported. When i was working on a bug fix yesterday, I had to update dependencies (probably just awesometypescriptloader) before webpack wouldn't complain, even though running tsc directly worked fine.

janriemer commented:
OK, I've rebased against master.
Everything looks fine now.

Thank you @chenkie and @escardin for your support.

janriemer commented:
Can this be merged now?

chenkie commented:
@escardin are any of the changes you're planning to make for the js-base64 issue going to impact this PR?

escardin commented:
I haven't got anything in progress, might as well merge this.
